### PR TITLE
Add protoc plugin for generating json schemas

### PR DESCRIPTION
**Pull request by alfredgunnar**
For some projects it is desired to get a JSON schema without having to use go.
With this plugin we can generate schemas as JSON files without first generating go code.

This also opens up the possibility of setting field descriptions in the schema from comments
in the proto file. That is not possible from generated go code.

### CODE REVIEW

File: encoding/jsonbq/marshal.go [hunk 1]
```diff
@@ -26,7 +26,19 @@ type Field struct {
 // The JSON conforms to the format in this example:
 // https://cloud.google.com/bigquery/docs/nested-repeated#example_schema
 func MarshalSchema(schema bigquery.Schema) ([]byte, error) {
+	return MarshalOptions{}.MarshalSchema(schema)
+}
+
+type MarshalOptions struct {
+	// Set to true to marshal json with indentation for better human readability
+	Indent bool
```
Comment (odsod) [inline encoding/jsonbq/marshal.go:34]: `Indent` should be a `string` to be consistent with the `protojson` API:

https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson#MarshalOptions

File: encoding/jsonbq/marshal.go [hunk 2]
```diff
@@ -26,7 +26,19 @@ type Field struct {
 // The JSON conforms to the format in this example:
 // https://cloud.google.com/bigquery/docs/nested-repeated#example_schema
 func MarshalSchema(schema bigquery.Schema) ([]byte, error) {
+	return MarshalOptions{}.MarshalSchema(schema)
+}
+
+type MarshalOptions struct {
```
Comment (odsod) [inline encoding/jsonbq/marshal.go:32]: missing GoDoc

File: encoding/jsonbq/marshal.go [hunk 3]
```diff
@@ -26,7 +26,19 @@ type Field struct {
...
+	// Set to true to marshal json with indentation for better human readability
+	Indent bool
+}
+
+func (o MarshalOptions) MarshalSchema(schema bigquery.Schema) ([]byte, error) {
```
Comment (odsod) [inline encoding/jsonbq/marshal.go:37]: missing GoDoc

File: internal/examples/proto/buf.gen.yaml [hunk 4]
```diff
@@ -9,6 +9,11 @@ managed:
 
 plugins:
   - name: go
-    out: gen
+    out: gen/go
     opt: module=go.einride.tech/protobuf-bigquery/internal/examples/proto/gen
     path: ./build/protoc-gen-go
+
+  - name: json_schemas
```
Comment (odsod) [inline internal/examples/proto/buf.gen.yaml:16]: the `name` should typically be the exact name of the protoc plugin, i.e. `bq-json-schema` in this case

File: internal/examples/proto/buf.gen.yaml [hunk 5]
```diff
@@ -9,6 +9,11 @@ managed:
...
+    out: gen/json
+    opt: indent=true
+    path: ./build/protoc-gen-bq-json-schema
```
Comment (odsod) [inline internal/examples/proto/buf.gen.yaml:19]: update editor config to add trailing newline

File: protoc-gen-bq-json-schema/genjson/config.go [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package genjson
+
+import "github.com/spf13/pflag"
+
+type Config struct {
+	Indent bool
```
Comment (odsod) [inline protoc-gen-bq-json-schema/genjson/config.go:6]: same comment as above - let's make `Indent` a `string`!

File: protoc-gen-bq-json-schema/genjson/run.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+func Run(gen *protogen.Plugin, config Config) error {
+	var files protoregistry.Files
+	for _, file := range gen.Files {
+		if err := files.RegisterFile(file.Desc); err != nil {
+			return err
+		}
+	}
+	for _, f := range gen.Files {
+		if f.Generate {
```
Comment (odsod) [inline protoc-gen-bq-json-schema/genjson/run.go:16]: I usually do `if !f.Generate` and a `continue` to keep the happy path on a lower indentation level - WDYT?

File: protoc-gen-bq-json-schema/genjson/run.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package genjson
+
+import (
+	"google.golang.org/protobuf/compiler/protogen"
+	"google.golang.org/protobuf/reflect/protoregistry"
+)
+
+func Run(gen *protogen.Plugin, config Config) error {
+	var files protoregistry.Files
```
Comment (odsod) [inline protoc-gen-bq-json-schema/genjson/run.go:9]: why do we need to create a `protoregistry.Files` here? it looks like we don't use it?

File: protoc-gen-bq-json-schema/genjson/config.go [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+}
+
+func (c *Config) AddToFlagSet(flags *pflag.FlagSet) {
```
Comment (odsod) [inline protoc-gen-bq-json-schema/genjson/config.go:9]: really like this way of managing flags - nice!

**Review by odsod — COMMENTED** [review]
Looks great! Just some stylistic comments

File: encoding/jsonbq/marshal.go (hunk 1 above)
Comment (alfredgunnar) [reply encoding/jsonbq/marshal.go:34]: Nice! Will change that.

File: internal/examples/proto/buf.gen.yaml (hunk 5 above)
Comment (alfredgunnar) [reply internal/examples/proto/buf.gen.yaml:19]: Ah! Well spotted. Got a new laptop this week and I'm missing my old Goland config. Will set that up and fix. 👍🏼

File: protoc-gen-bq-json-schema/genjson/run.go (hunk 8 above)
Comment (alfredgunnar) [reply protoc-gen-bq-json-schema/genjson/run.go:9]: We don't. Copy-paste leftovers. Removing.

File: protoc-gen-bq-json-schema/genjson/config.go (hunk 9 above)
Comment (alfredgunnar) [reply protoc-gen-bq-json-schema/genjson/config.go:9]: Yeah, it's a copy-paste from https://github.com/einride/protoc-gen-go-cli 😉

File: protoc-gen-bq-json-schema/genjson/run.go (hunk 7 above)
Comment (alfredgunnar) [reply protoc-gen-bq-json-schema/genjson/run.go:16]: Sure! Changing.

**Comment by alfredgunnar** [comment]
@odsod I just found out that [this](https://github.com/GoogleCloudPlatform/protoc-gen-bq-schema) exists. But that requires copying/depending on the `.proto` files from the root of that repo.

I wonder if it maybe makes sense to still have our own version after all? What do you think?

File: protoc-gen-bq-json-schema/genjson/config.go (hunk 6 above)
Comment (alfredgunnar) [reply protoc-gen-bq-json-schema/genjson/config.go]: I had some issues with providing a string with just spaces as an `opt` in `buf.gen.yaml`. So I changed this to be called `Format` instead to make a distinction from the `Indent` options in `jsonbq.MarshalOptions` which is now a string.

**Comment by alethenorio** [comment]
I think this looks good but if it maps 1 to 1 with the google plugin it might be better to keep that one? What do you think?

**Comment by alfredgunnar** [comment]
> I think this looks good but if it maps 1 to 1 with the google plugin it might be better to keep that one? What do you think @alfredgunnar ?

@alethenorio I think there are good things in both of them. I think we have some nice options [here](https://github.com/einride/protobuf-bigquery-go/blob/f7813beb1b8e63e29ce58fe075968e16e15dc498/encoding/protobq/schema.go#L19) for schema generation. But the google plugin uses proto annotations (which is rather nice but requires a dependency on those proto files) for some field settings.

**Comment by alethenorio** [comment]
I think you have very good point there. If you think the PR is ready for review, make it ready for review and I'll be happy to approve the PR

**Review by alethenorio — APPROVED** [review]

**Pull request merged**